lib/Connection.js

```diff
@@ -340,5 +360,24 @@ function prepareHeaders (headers = {}, auth) {
   return headers
 }
 
+function getIssuerCertificate (socket) {
+  let certificate = socket.getPeerCertificate(true)
+  while (certificate && Object.keys(certificate).length > 0) {
+    /* istanbul ignore else */
+    if (certificate.issuerCertificate !== undefined) {
+      // For self-signed certificates, `issuerCertificate` may be a circular reference.
+      /* istanbul ignore else */
+      if (certificate.fingerprint256 === certificate.issuerCertificate.fingerprint256) {
+        break
+      }
+      /* istanbul ignore next */
+      certificate = certificate.issuerCertificate
+    } else {
+      break
```

Comment:
> when issuerCertificate is undefined, the certificate is invalid or malformed, let's not try to use it.

Hmm, it seems that missing `issuerCertificate` in certificate is one thing and missing `issuerCertificate` property in a certificate parsed by NodeJS is a different thing. If host just doesn't return a full chain, I _guess_, we'll have an empty `issuerCertificate` property (or if we pass `false` to `socket.getPeerCertificate`).

If ES is configured to not return full chain do we treat it as a misconfiguration @jkakavas or you meant something else?

Reply:
>  missing issuerCertificate in certificate is one thing and missing issuerCertificate property in a certificate parsed by NodeJS is a different thing

Correct. I focused on the former I guess :/ Still though, we're in a method called `getIssuerCertificate` and we shouldn't return a certificate just because we couldn't get it's issuer. If ES doesn't return the full chain, then `caFingerprint` does not make much sense to be set.

Reply:
> Still though, we're in a method called getIssuerCertificate and we shouldn't return a certificate just because we couldn't get it's issuer.

Yep

>  If ES doesn't return the full chain, then `caFingerprint` does not make much sense to be set.

Would it make sense to rename it to `rootCAFingerprint` then? Because we may still have a number intermediate CA certificates in the chain.

Reply:
> Would it make sense to rename it to rootCAFingerprint then? Because we may still have a number intermediate CA certificates in the chain even though root CA certificate isn't there.

Possibly, it makes it clearer I think but we can handle this with documentation also. 

> And just for my own education, if I issue a Let's Encrypt certificate to use with my ES instance, I still have to configure ES with the full chain, even though root CA certificate is issued by the well-known public CA, right?

Well, we don't support the enrollment process for arbitrary configurations but if you'd want to use let's encrypt and use `caFingerprint` in a way in your es client, then yes ,you'd need to do that. 

Generically, it would make more sense for a client to support leaf certificate pinning ( as in I want to trust _this_ server certificate only ). The idea behind us using CA certificate pinning for the enrollment process is that it allows you to get a single enrollment token that might contain many addresses for all ES nodes in a cluster and you can try/selct any of them and validate the connection with a single fingerprint ( as opposed to one per node )

Reply:
Yeah, but I had an impression that `caFingerprint` we've introduced here is not specific to enrollment process otherwise we should have named it differently I guess.

> Generically, it would make more sense for a client to support leaf certificate pinning ( as in I want to trust this server certificate only ).

That's actually what I was alluding to initially, pinning to the leaf (this specific server) or intermediate CA (this specific group of servers) seems to be a perfectly valid use case. My concern was that ES seems to not explicitly require full chains and moreover works fine with non-complete chains. That'd be enough for the aforementioned use case, but it won't work if we treat certificate with null/undefined `issuerCertificate` as invalid (the one that NodeJS parses, aka `if (certificate.issuerCertificate == null) return null`). 

Anyway, PR is merged, so let's if it ever becomes a problem :man_shrugging:

Reply:
> That's actually what I was alluding to initially, pinning to the leaf (this specific server) or intermediate CA (this specific group of servers) seems to be a perfectly valid use case. 

Good point, I'd be :+1: This would cover security on by default enrollment process, and more generic cases at the same time. Maybe we can do in a followup ?

Reply:
> Maybe we can do in a followup ?

Yeah, I'd :+1: too.

Reply:
I'm exactly having this issue. Returned certificate only has `issuer` not `issuerCertificate`. It results in `error: Unhandled Rejection at: Promise [object Promise] reason ConnectionError: Invalid or malformed certificate`. This is in contrast with Python client where `ssl_assert_fingerprint` parameter works as expected with self-signed certificate that elastic generates by default.